File: javascript/src/resources/companies.ts
/** Companies House company, officer, PSC, charge and filing data. */

import { seg, type QueryParams, type Transport } from '../transport.js';
import type { JsonObject } from '../types.js';

/** Options accepted by every call: cancellation via an `AbortSignal`. */
export interface CallOptions {
  signal?: AbortSignal;
}

export interface PageOptions extends CallOptions {
  /** Page size. The API applies its own default and cap. */
  itemsPerPage?: number;
  /** Zero-based offset for paging. */
  startIndex?: number;
}

export interface FilingHistoryOptions extends PageOptions {
  /** Companies House filing category, e.g. `"accounts"`, `"confirmation-statement"`. */
  category?: string;
}

/** Filters for {@link Companies.advancedSearch}. Dates are `YYYY-MM-DD`. */
export interface AdvancedSearchOptions extends PageOptions {
  companyNameIncludes?: string;
  companyNameExcludes?: string;
  /** Repeatable: `["active", "liquidation"]` matches either. */
  companyStatus?: string[];
  companyType?: string[];
  location?: string;
  sicCodes?: string[];
  incorporatedFrom?: string;
  incorporatedTo?: string;
  dissolvedFrom?: string;
  dissolvedTo?: string;
  size?: number;
}

/**
 * Company registry lookups. Available on every plan.
 *
 * Company numbers are Companies House registration numbers — up to eight
 * alphanumeric characters, e.g. `"00445790"` or `"SC123456"`. Leading zeroes are
 * significant, so keep them as strings.
 */
export class Companies {
  constructor(private readonly transport: Transport) {}

  /** Search companies by name or number. */
  async search(q: string, options: PageOptions = {}): Promise<JsonObject> {
    return this.transport.get('/companies/search', {
      params: { q, items_per_page: options.itemsPerPage, start_index: options.startIndex },
      signal: options.signal,
    });
  }

  /** Search officers (directors, secretaries) by name across all companies. */
  async searchOfficers(q: string, options: PageOptions = {}): Promise<JsonObject> {
    return this.transport.get('/companies/search/officers', {
      params: { q, items_per_page: options.itemsPerPage, start_index: options.startIndex },
      signal: options.signal,
    });
  }

  /** Structured company search with status, type, SIC and date filters. */
  async advancedSearch(options: AdvancedSearchOptions = {}): Promise<JsonObject> {
    const params: QueryParams = {
      company_name_includes: options.companyNameIncludes,
      company_name_excludes: options.companyNameExcludes,
      company_status: options.companyStatus,
      company_type: options.companyType,
      location: options.location,
      sic_codes: options.sicCodes,
      incorporated_from: options.incorporatedFrom,
      incorporated_to: options.incorporatedTo,
      dissolved_from: options.dissolvedFrom,
      dissolved_to: options.dissolvedTo,
      size: options.size,
      start_index: options.startIndex,
      items_per_page: options.itemsPerPage,
    };
    return this.transport.get('/companies/advanced-search', { params, signal: options.signal });
  }

  /** The company profile: name, status, type, dates, address, SIC codes. */
  async get(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(`/companies/${seg(companyNumber, 'companyNumber')}`, options);
  }

  /**
   * Profile, officers, PSCs, charges, insolvency and filings in one call.
   *
   * Cheaper than issuing each lookup separately — one request against your rate
   * limit instead of six.
   */
  async dossier(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(`/companies/${seg(companyNumber, 'companyNumber')}/dossier`, options);
  }

  /** Appointed officers, including resigned ones. */
  async officers(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/officers`,
      options,
    );
  }

  /** Registered beneficial owners (PSCs). */
  async pscs(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/persons-with-significant-control`,
      options,
    );
  }

  /** PSC statements — the register's explanations for an absent PSC. */
  async pscStatements(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/persons-with-significant-control-statements`,
      options,
    );
  }

  /** How many corporate layers sit between the company and a real person. */
  async pscChainDepth(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/psc-chain-depth`,
      options,
    );
  }

  /** The full ownership chain as a tree, resolving corporate PSCs upwards. */
  async pscChainTree(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/psc-chain-tree`,
      options,
    );
  }

  /** Registered charges (secured debt) against the company. */
  async charges(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(`/companies/${seg(companyNumber, 'companyNumber')}/charges`, options);
  }

  /** One charge in full, including the persons entitled to it. */
  async charge(
    companyNumber: string,
    chargeId: string,
    options: CallOptions = {},
  ): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/charges/${seg(chargeId, 'chargeId')}`,
      options,
    );
  }

  /** Insolvency cases. Returns an empty payload when there are none. */
  async insolvency(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/insolvency`,
      options,
    );
  }

  /** Disqualified-director records across the company's officers. */
  async disqualifications(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/disqualifications`,
      options,
    );
  }

  /** Disqualification record for one officer of this company. */
  async officerDisqualification(
    companyNumber: string,
    officerId: string,
    options: CallOptions = {},
  ): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/officers/${seg(officerId, 'officerId')}/disqualification`,
      options,
    );
  }

  /** Every appointment held by one officer, across all companies. */
  async officerAppointments(officerId: string, options: PageOptions = {}): Promise<JsonObject> {
    return this.transport.get(`/companies/officers/${seg(officerId, 'officerId')}/appointments`, {
      params: { items_per_page: options.itemsPerPage, start_index: options.startIndex },
      signal: options.signal,
    });
  }

  // ── Professional plan ──────────────────────────────────────────────────────

  /** Filing history. **Requires an active Professional subscription.** */
  async filingHistory(
    companyNumber: string,
    options: FilingHistoryOptions = {},
  ): Promise<JsonObject> {
    return this.transport.get(`/companies/${seg(companyNumber, 'companyNumber')}/filing-history`, {
      params: {
        start_index: options.startIndex,
        items_per_page: options.itemsPerPage,
        category: options.category,
      },
      signal: options.signal,
    });
  }

  /**
   * Structured data parsed out of one filing.
   *
   * **Requires an active Professional subscription.**
   */
  async filingExtract(
    companyNumber: string,
    transactionId: string,
    options: CallOptions = {},
  ): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/filing-history/${seg(transactionId, 'transactionId')}/extract`,
      options,
    );
  }

  /**
   * Share capital and shareholders from the latest confirmation statement.
   *
   * **Requires an active Professional subscription.**
   */
  async statementOfCapital(companyNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/companies/${seg(companyNumber, 'companyNumber')}/statement-of-capital`,
      options,
    );
  }
}

File: elixir/lib/kyccentral/companies.ex
defmodule KYCCentral.Companies do
  @moduledoc """
  Companies House company, officer, PSC, charge and filing data.

  Available on every plan, except the three endpoints marked
  **Professional plan**.

  Company numbers are Companies House registration numbers — up to eight
  alphanumeric characters, e.g. `"00445790"` or `"SC123456"`. Leading zeroes are
  significant, so keep them as strings.
  """

  import KYCCentral.Transport, only: [segment: 2]

  alias KYCCentral.{Error, Transport}

  @type result :: {:ok, map()} | {:error, Error.t()}

  @doc """
  Search companies by name or number.

  ## Options

    * `:items_per_page` — page size. The API applies its own default and cap.
    * `:start_index` — zero-based offset for paging.
  """
  @spec search(KYCCentral.t(), String.t(), keyword()) :: result()
  def search(client, q, opts \\ []) do
    Transport.request(client, :get, "/companies/search", params: page_params(q, opts))
  end

  @doc "Search officers (directors, secretaries) by name across all companies."
  @spec search_officers(KYCCentral.t(), String.t(), keyword()) :: result()
  def search_officers(client, q, opts \\ []) do
    Transport.request(client, :get, "/companies/search/officers", params: page_params(q, opts))
  end

  @doc """
  Structured company search with status, type, SIC and date filters.

  ## Options

  All optional. Dates are `YYYY-MM-DD`. List options are repeatable, so
  `company_status: ["active", "liquidation"]` matches either.

    * `:company_name_includes`, `:company_name_excludes`
    * `:company_status`, `:company_type`, `:sic_codes` — lists
    * `:location`
    * `:incorporated_from`, `:incorporated_to`
    * `:dissolved_from`, `:dissolved_to`
    * `:size`, `:start_index`, `:items_per_page`
  """
  @spec advanced_search(KYCCentral.t(), keyword()) :: result()
  def advanced_search(client, opts \\ []) do
    params = [
      company_name_includes: opts[:company_name_includes],
      company_name_excludes: opts[:company_name_excludes],
      company_status: opts[:company_status],
      company_type: opts[:company_type],
      location: opts[:location],
      sic_codes: opts[:sic_codes],
      incorporated_from: opts[:incorporated_from],
      incorporated_to: opts[:incorporated_to],
      dissolved_from: opts[:dissolved_from],
      dissolved_to: opts[:dissolved_to],
      size: opts[:size],
      start_index: opts[:start_index],
      items_per_page: opts[:items_per_page]
    ]

    Transport.request(client, :get, "/companies/advanced-search", params: params)
  end

  @doc "The company profile: name, status, type, dates, address, SIC codes."
  @spec get(KYCCentral.t(), String.t()) :: result()
  def get(client, company_number), do: company_path(client, company_number, "")

  @doc """
  Profile, officers, PSCs, charges, insolvency and filings in one call.

  Cheaper than issuing each lookup separately — one request against your rate
  limit instead of six.
  """
  @spec dossier(KYCCentral.t(), String.t()) :: result()
  def dossier(client, company_number), do: company_path(client, company_number, "/dossier")

  @doc "Appointed officers, including resigned ones."
  @spec officers(KYCCentral.t(), String.t()) :: result()
  def officers(client, company_number), do: company_path(client, company_number, "/officers")

  @doc "Registered beneficial owners (PSCs)."
  @spec pscs(KYCCentral.t(), String.t()) :: result()
  def pscs(client, company_number) do
    company_path(client, company_number, "/persons-with-significant-control")
  end

  @doc "PSC statements — the register's explanations for an absent PSC."
  @spec psc_statements(KYCCentral.t(), String.t()) :: result()
  def psc_statements(client, company_number) do
    company_path(client, company_number, "/persons-with-significant-control-statements")
  end

  @doc "How many corporate layers sit between the company and a real person."
  @spec psc_chain_depth(KYCCentral.t(), String.t()) :: result()
  def psc_chain_depth(client, company_number) do
    company_path(client, company_number, "/psc-chain-depth")
  end

  @doc "The full ownership chain as a tree, resolving corporate PSCs upwards."
  @spec psc_chain_tree(KYCCentral.t(), String.t()) :: result()
  def psc_chain_tree(client, company_number) do
    company_path(client, company_number, "/psc-chain-tree")
  end

  @doc "Registered charges (secured debt) against the company."
  @spec charges(KYCCentral.t(), String.t()) :: result()
  def charges(client, company_number), do: company_path(client, company_number, "/charges")

  @doc "One charge in full, including the persons entitled to it."
  @spec charge(KYCCentral.t(), String.t(), String.t()) :: result()
  def charge(client, company_number, charge_id) do
    with {:ok, number} <- segment(company_number, "company_number"),
         {:ok, id} <- segment(charge_id, "charge_id") do
      Transport.request(client, :get, "/companies/#{number}/charges/#{id}")
    end
  end

  @doc "Insolvency cases. Returns an empty payload when there are none."
  @spec insolvency(KYCCentral.t(), String.t()) :: result()
  def insolvency(client, company_number), do: company_path(client, company_number, "/insolvency")

  @doc "Disqualified-director records across the company's officers."
  @spec disqualifications(KYCCentral.t(), String.t()) :: result()
  def disqualifications(client, company_number) do
    company_path(client, company_number, "/disqualifications")
  end

  @doc "Disqualification record for one officer of this company."
  @spec officer_disqualification(KYCCentral.t(), String.t(), String.t()) :: result()
  def officer_disqualification(client, company_number, officer_id) do
    with {:ok, number} <- segment(company_number, "company_number"),
         {:ok, officer} <- segment(officer_id, "officer_id") do
      Transport.request(
        client,
        :get,
        "/companies/#{number}/officers/#{officer}/disqualification"
      )
    end
  end

  @doc "Every appointment held by one officer, across all companies."
  @spec officer_appointments(KYCCentral.t(), String.t(), keyword()) :: result()
  def officer_appointments(client, officer_id, opts \\ []) do
    with {:ok, officer} <- segment(officer_id, "officer_id") do
      params = [items_per_page: opts[:items_per_page], start_index: opts[:start_index]]

      Transport.request(client, :get, "/companies/officers/#{officer}/appointments",
        params: params
      )
    end
  end

  @doc """
  Filing history. **Requires an active Professional subscription.**

  ## Options

    * `:category` — Companies House filing category, e.g. `"accounts"`,
      `"confirmation-statement"`, `"officers"`.
    * `:items_per_page`, `:start_index`
  """
  @spec filing_history(KYCCentral.t(), String.t(), keyword()) :: result()
  def filing_history(client, company_number, opts \\ []) do
    with {:ok, number} <- segment(company_number, "company_number") do
      params = [
        start_index: opts[:start_index],
        items_per_page: opts[:items_per_page],
        category: opts[:category]
      ]

      Transport.request(client, :get, "/companies/#{number}/filing-history", params: params)
    end
  end

  @doc """
  Structured data parsed out of one filing.

  **Requires an active Professional subscription.**
  """
  @spec filing_extract(KYCCentral.t(), String.t(), String.t()) :: result()
  def filing_extract(client, company_number, transaction_id) do
    with {:ok, number} <- segment(company_number, "company_number"),
         {:ok, transaction} <- segment(transaction_id, "transaction_id") do
      Transport.request(
        client,
        :get,
        "/companies/#{number}/filing-history/#{transaction}/extract"
      )
    end
  end

  @doc """
  Share capital and shareholders from the latest confirmation statement.

  **Requires an active Professional subscription.**
  """
  @spec statement_of_capital(KYCCentral.t(), String.t()) :: result()
  def statement_of_capital(client, company_number) do
    company_path(client, company_number, "/statement-of-capital")
  end

  defp company_path(client, company_number, suffix) do
    with {:ok, number} <- segment(company_number, "company_number") do
      Transport.request(client, :get, "/companies/#{number}#{suffix}")
    end
  end

  defp page_params(q, opts) do
    [q: q, items_per_page: opts[:items_per_page], start_index: opts[:start_index]]
  end
end
